### PR TITLE
samples: nrf9160: mqtt_simple: Add extended TLS support

### PR DESCRIPTION
**Pull request by simensrostad**
Add certificate provisioning, sec tag socket configuration
and TLS session caching option.

### CODE REVIEW

File: samples/nrf9160/mqtt_simple/Kconfig [hunk 1]
```diff
@@ -33,8 +33,19 @@ config MQTT_PAYLOAD_BUFFER_SIZE
...
+
+config MQTT_TLS_PEER_VERIFY
+	int "Set peer verification level"
+	default 2
+
 endmenu
 
 menu "Zephyr Kernel"
-source "$ZEPHYR_BASE/Kconfig.zephyr"
+source "Kconfig.zephyr"
```
Comment (sigvartmh) [inline samples/nrf9160/mqtt_simple/Kconfig:50]: Why this change? I don't think it should be in here.

Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/Kconfig:50]: The NordicBuilder suggested this change.

Comment (sigvartmh) [reply samples/nrf9160/mqtt_simple/Kconfig:50]: Then I guess it's fine :+1: just looked like a weird change. I guess it's because the Kconfig of the sample is quite old.

**Review by sigvartmh — APPROVED** [review]
LGTM

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 2]
```diff
@@ -332,10 +381,39 @@ static void client_init(struct mqtt_client *client)
...
 	client->transport.type = MQTT_TRANSPORT_SECURE;
+
+#if defined(CONFIG_BSD_LIBRARY)
+	static sec_tag_t sec_tag_list[] = { CONFIG_MQTT_TLS_SEC_TAG };
+
+	tls_cfg->session_cache = IS_ENABLED(CONFIG_MQTT_TLS_SESSION_CACHING) ?
+					    TLS_SESSION_CACHE_ENABLED :
+					    TLS_SESSION_CACHE_DISABLED;
+
+	tls_cfg->peer_verify = CONFIG_MQTT_TLS_PEER_VERIFY;
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/src/main.c:394]: Isn't it needed to set these parameters for qemu build as well? Right now running the sample for qemu will skip the certificate verification if I understand correctly.

File: samples/nrf9160/mqtt_simple/src/certificates.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#define CLIENT_PRIVATE_KEY \
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/src/certificates.h:7]: Do we really need to authenticate the client in this simple sample? Wouldn't server certificate verification be enough? I'm not even sure if the default server (mqtt.eclipse.org) allows that.

**Review by rlubos — COMMENTED** [review]
What server was used for testing? Perhaps we should provide the certificate for the default server (mqtt.eclipse.org, from what I've seen it supports TLS)., so that the sample can be run with TLS out of the box? We should also add an overlay file to enable TLS for nRF91 (and possibly mentions TLS support in the sample documentation).

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 2 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c:376]: Yes, you are correct.

File: samples/nrf9160/mqtt_simple/src/certificates.h (hunk 3 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/certificates.h]: Yes, the server certificate should be enough. I have added the CA cert for "test.mosquitto.org".

**Comment by simensrostad** [comment]
@rlubos I have addressed your review comments. I have done some testing and the sample now works out of the box with no TLS/TLS for both qemu and the nrf91.

File: samples/nrf9160/mqtt_simple/CMakeLists.txt [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018 Nordic Semiconductor
+# Copyright (c) 2020 Nordic Semiconductor
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/CMakeLists.txt:2]: There's no need to update the year here, there were no modifications in this file.

File: samples/nrf9160/mqtt_simple/overlay-carrier.conf [hunk 5]
```diff
@@ -1,8 +1,6 @@
-#
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/overlay-carrier.conf:1]: Let's keep the copyright notice.

File: samples/nrf9160/mqtt_simple/overlay-qemu-x86-tls.conf [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# To build the mqtt simple sample with TLS support
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/overlay-qemu-x86-tls.conf:1]: I think this information should rather be placed in the README.rst, let's add the copyright info instead.

File: samples/nrf9160/mqtt_simple/overlay-tls.conf [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# To build the mqtt simple sample with TLS support
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/overlay-tls.conf:1]: Same story here, please add copyright notice.

The README.rst has already a section on how to add TLS support to the sample manually, which IMO is no longer required after this update. Perhaps we should replace it with information about overlays:
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/nrf9160/mqtt_simple/README.rst#references

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 8]
```diff
@@ -332,10 +368,38 @@ static void client_init(struct mqtt_client *client)
...
+
+#if defined(CONFIG_BSD_LIBRARY)
+	tls_cfg->session_cache = IS_ENABLED(CONFIG_MQTT_TLS_SESSION_CACHING) ?
+					    TLS_SESSION_CACHE_ENABLED :
+					    TLS_SESSION_CACHE_DISABLED;
+#else
+	/* TLS session caching is not supported by the Zephyr network stack */
+	tls_cfg->session_cache = TLS_SESSION_CACHE_DISABLED;
+
+	err = certificates_provision();
```
Comment (rlubos) [inline samples/nrf9160/mqtt_simple/src/main.c:391]: If I understand correctly, the user will need to provision certificate to the modem manually, before running the sample. Perhaps we could automate the process, just as `https_client` sample does, and pre-load the certificate into the modem in the sample?
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/nrf9160/https_client/src/main.c#L171
Note, that certificate should be loaded before `modem_configure` (autoconnect is already disabled in this sample).

**Review by rlubos — COMMENTED** [review]
Looks pretty good, just some suggestions to make the process simpler for the users.

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 9]
```diff
@@ -395,7 +459,10 @@ void main(void)
 
 	modem_configure();
 
-	client_init(&client);
+	err = client_init(&client);
+	if (err != 0) {
+		printk("ERROR: client_init %d\n", err);
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/src/main.c:464]: It looks like an error here could cause the subsequent functions to fail (depending on peer verification settings) so it should be considered to return.

Slightly besides the scope of this PR: I think it should also be considered to let `broker_init()` return an error and propagate it from `client_init()`. Error handling in `broker_init()` looks incomplete, and failure in that function would cause the connection attempt to fail.

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 10]
```diff
@@ -34,6 +38,31 @@ static bool connected;
...
+	err = tls_credential_add(CONFIG_MQTT_TLS_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+				 CA_CERTIFICATE,
+				 sizeof(CA_CERTIFICATE));
+	if (err < 0) {
+		printk("Failed to register CA certificate: %d\n", err);
+		return err;
+	}
+
+	certs_added = true;
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/src/main.c:60]: Looks like this function is only called once, so the flag is sort of redundant in this use-case

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 11]
```diff
@@ -10,6 +10,10 @@
 #include <string.h>
 #include <random/rand32.h>
 
+#if defined(CONFIG_BOARD_QEMU_X86) && !defined(CONFIG_BSD_LIBRARY)
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/src/main.c:13]: Let's include this file for all builds, and offer certificate provisioning functionality also for builds using BSDlib

File: samples/nrf9160/mqtt_simple/overlay-tls.conf [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+# add the following to your west build command:
+# -- -DOVERLAY_CONFIG=overlay-tls.conf
+
+# MQTT TLS
+CONFIG_MQTT_LIB_TLS=y
+CONFIG_MQTT_BROKER_HOSTNAME="test.mosquitto.org"
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/overlay-tls.conf:7]: I tested that it's also possible to connect to the default broker `mqtt.eclipse.org`

**Review by jtguggedal — COMMENTED** [review]
Overall looks good! Left a few comments to consider.

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 8 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c:391]: You are correct. Your suggestion sounds like a fine improvement. I will add certificate provisioning for the nrf91.

File: samples/nrf9160/mqtt_simple/overlay-tls.conf (hunk 12 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/overlay-tls.conf:7]: Very good, without changing the CA?

**Comment by simensrostad** [comment]
@umapraseeda I will update the documentation for this sample. It will need a documentation review, thanks.

File: samples/nrf9160/mqtt_simple/overlay-tls.conf (hunk 7 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/overlay-tls.conf:1]: Sound good. I will update the documentation to reflect the changes that have been made, and remove/rephrase how the external link regarding TLS support is added. It makes sense to link to the external TLS doc for potential issues related to TLS. If people have issues they can post comments there.

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 9 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c:464]: Let's add this while we are at it.

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 10 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c]: Removed the flag

File: samples/nrf9160/mqtt_simple/overlay-tls.conf (hunk 12 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/overlay-tls.conf]: Changed the default broker to mqtt.eclipse.org and added the corresponding CA.

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 13]
```diff
@@ -305,15 +346,22 @@ static void broker_init(void)
...
 {
+	int err;
+
 	mqtt_client_init(client);
 
-	broker_init();
+	err = broker_init();
+	if (err) {
+		printk("Failed to initialize broker connection\n");
+	}
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/src/main.c:364]: Error should be returned here

File: samples/nrf9160/mqtt_simple/src/main.c [hunk 14]
```diff
@@ -266,12 +310,10 @@ static void broker_init(void)
 	err = getaddrinfo(CONFIG_MQTT_BROKER_HOSTNAME, NULL, &hints, &result);
 	if (err) {
 		printk("ERROR: getaddrinfo failed %d\n", err);
-
-		return;
+		return err;
```
Comment (jtguggedal) [inline samples/nrf9160/mqtt_simple/src/main.c:313]: Note that `getaddrinfo` returns positive error codes

Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c]: Changed to -err;

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 13 above)
Comment (simensrostad) [reply samples/nrf9160/mqtt_simple/src/main.c:365]: woops, added

File: samples/nrf9160/mqtt_simple/src/main.c (hunk 14 above)
Comment (jtguggedal) [reply samples/nrf9160/mqtt_simple/src/main.c]: Okay, that's confusing. Better translate it to a negative error code than (as I see you've already done)

**Review by rlubos — APPROVED** [review]
LGTM, thanks!

**Review by jtguggedal — APPROVED** [review]
LGTM

**Review by umapraseeda — APPROVED** [review]
